feat: Replace nginx snippet annotation with custom header annotation

HerrSpeck opened the pull request:
Our cluster does not allow snippet annotations by default and changing the configuration would be complicated.

But the snippet is not actually needed, since the same thing can be achieve using a config map and the custom-headers annotation.

santiagon610 commented:
Does this functionality hold harmless users that do not require this functionality?

If so, LGTM. Can you increment the version in `Chart.yaml`?

santiagon610 commented on charts/vaultwarden/templates/ingress.yaml:
```diff
@@ -1,3 +1,10 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: customer-headers-configmap
+data:
+  Request-Id: $req_id
```
Does this functionality hold harmless users that do not require this functionality, or should it also be placed behind an `{{- if $ingress.nginxIngressAnnotations }}` flag?

HerrSpeck replied:
Since this just defines a config map that is never referenced anywhere if the nginxIngressAnnotations variable is disabled, my guess would be that we wouldn't really have to hide it behind the flag.

That said, I'm more or less just getting started with kubernetes, so my guess might be off here. If I do add the flag, would I have to hide the separator (---) behind the flag as well?

guerzon replied:
`ConfigMap` definitions should be added to `charts/vaultwarden/templates/configmap.yaml`.

Also, there is a typo in the name (should be "custom-headers-configmap not "customer-headers-configmap").

santiagon610 commented on charts/vaultwarden/templates/ingress.yaml:
```diff
@@ -26,8 +33,7 @@ metadata:
     {{- if $ingress.nginxIngressAnnotations }}
-    nginx.ingress.kubernetes.io/configuration-snippet: |
-      more_set_headers "Request-Id: $req_id";
+    nginx.ingress.kubernetes.io/custom-headers: custom-headers-configmap
```
Should this be parameterized so that the ConfigMap can be called anything, but defaulted in `values.yaml`?

HerrSpeck replied:
Done :) I added a new variable named $ingress.customHeaderConfigMap that can be used to define a custom config map name in case someone wants to override it and it defaults to the newly added config map.

santiagon610 reviewed (commented):
Overall, looks good. I would just worry about breaking functionality for other users that are using Nginx Ingress with your changes.

Also, since this is a feature release, could you update the `Chart.yaml` to increment the minor version of the release?

HerrSpeck commented:
> Overall, looks good. I would just worry about breaking functionality for other users that are using Nginx Ingress with your changes.
> 
> Also, since this is a feature release, could you update the `Chart.yaml` to increment the minor version of the release?

Changed the version :)
I'm not sure if it would interfere with other users. There was no way before to set any custom headers before, so I'm not overriding anything. Also, the Request-Id header is still only added if the nginxIngressAnnotations are activated, which is the same behavior as before.

guerzon commented:
Hallo @HerrSpeck, danke für die PR

I am adding some review points.

guerzon commented on charts/vaultwarden/templates/ingress.yaml:
```diff
@@ -1,3 +1,10 @@
+metadata:
+  name: customer-headers-configmap
+data:
+  Request-Id: $req_id
```
According to the documentation [here](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#custom-headers), you should also add `global-allowed-response-headers` to the ConfigMap.

This is the example I found:

```
global-allowed-response-headers: Referrer-Policy,Strict-Transport-Security,X-Content-Type-Options,X-Frame-Options,Content-Security-Policy
```

HerrSpeck commented:
Thanks for the response! :)

You're totally right, I moved the config map to the configmap.yaml file :) and I also fixed the typo, not sure how that happened 😅 

I did some digging into the global-allowed-response-headers flag and from what I can tell this is not actually a key in the custom-headers config map, but is instead defined in the nginx-configuration config map. As far as I can tell, it controls which headers are allowed in general and as soon as you would like to set custom headers, you also have to specify this value and "whitelist" the custom headers.

So I added another config map for the nginx-configuration and I put it in the same namespace as the ingress configuration. Kind of lost at this point though, because I'm not sure if that's enough for the value to be applied or if we still have to add a reference to this config map somewhere. My current understanding is that this should suffice, but maybe you can help me out if I do in fact have to reference it somewhere else :)

HerrSpeck commented:
Hi there @guerzon !

I had a colleague take a look at the problem as well and we found some more information that I would like to share:

It looks like this is a flag/setting on the ingress-controller and I couldn't find any way to configure this using these charts at all. Instead, as the name suggests, this is a global configuration (like the allow-snippet-annotations flag that caused this whole issue on our side) that has to be applied directly to the ingress-controller. Therefore, I reverted the changes where I introduced an additional config map that I tried to use to allow the correct headers.

BUT: I also found out that the [allow-snippet-annotations configuration is deprecated](https://github.com/kubernetes/kubernetes/issues/126811) and should not be used, because of security issues.
Therefore, I still think this is the correct way to solve this issue and I would suggest that we add a notice/warning to the README, that tells readers to configure the global-allowed-response-headers field of their ingress-controller to allow the Request-Id header. What do you think?

guerzon commented:
Thanks for the feedback @HerrSpeck. I had more time to test this and I can see no issue with not adding `global-allowed-response-headers`.

PR looks good and ready to merge, but please rebase and fix the merge conflict. Thanks.

HerrSpeck commented:
@guerzon Alright, sounds great :) I rebased the pr and fixed the merge conflicts. Setting 0.25.0 as the new version is alright? Or should I just increase the patch version instead?

RealKelsar commented:
This broke my Setup. Not exactly sure why yet. Gives me 503 now, but other then in the nginx log I can not see any errors. Need to investigate after work.

guerzon commented:
> This broke my Setup. Not exactly sure why yet. Gives me 503 now, but other then in the nginx log I can not see any errors. Need to investigate after work.

Hi @RealKelsar, please submit a [new issue](https://github.com/guerzon/vaultwarden/issues/new) and include your config and logs. Thanks.
